### PR TITLE
Added ability to "Upgrade" a tool

### PR DESCRIPTION
**Pull request by xoen**
Similarly to how a user can deploy an instance of a tool, this will allow users to upgrade a tool (the `helm` command is the same).

We only show the "Upgrade" button when a user's deployed tool version differs from the version in the DB (which is the version we consider stable and supported).

Also:
- fixed "Restart" button style/layout
- fixed `jsConfirm` module to work with buttons too (e.g. the "delete data source" button was a `<button>` with `.js-confirm` class but it didn't ask for confirmation)
- removed unused `delete_button` jinja macro
- commented redis server pidfile dependency as it doesn't seem to work on my machine (and probably on other platforms too)

**Ticket**: https://trello.com/c/rz6o5bug/101-users-can-upgrade-their-tools

### CODE REVIEW

File: controlpanel/api/models/tool.py [hunk 1]
```diff
@@ -49,17 +49,24 @@ def create(self, *args, **kwargs):
...
-                Q(chart_name=deployment.metadata.labels["app"])
+                Q(chart_name=chart_name)
                 # & Q(version=version)
             )
 
         tools = Tool.objects.filter(filter)
-        return [ToolDeployment(tool, user) for tool in tools]
+        results = []
+        for tool in tools:
+            outdated = tool.version != deployed_versions[tool.chart_name]
```
Comment (andyhd) [inline controlpanel/api/models/tool.py]: Not too comfortable with `!=`, although I experimented with a couple of semver python modules and they're all a bit unpleasant, so stick with this (at least for now)

File: controlpanel/api/models/tool.py [hunk 2]
```diff
@@ -73,11 +80,12 @@ class ToolDeployment:
 
     objects = ToolDeploymentManager()
 
-    def __init__(self, tool, user):
+    def __init__(self, tool, user, outdated=False):
         self._id_token = None
         self._subprocess = None
         self.tool = tool
         self.user = user
+        self.outdated = outdated
```
Comment (andyhd) [inline controlpanel/api/models/tool.py]: I think `outdated` would be better as a property or `is_outdated` method on `ToolDeployment`. Maybe pass in the `deployed_version` to the constructor instead?

File: controlpanel/frontend/consumers.py [hunk 3]
```diff
@@ -126,6 +128,15 @@ def tool_deploy(self, message):
...
             log.debug(f"Deployed {tool.name} for {user}")
 
+    def tool_upgrade(self, message):
+        """
+        Upgrade simply means re-installing the helm chart for the tool
+        """
+        self.tool_deploy(message)
+
+        tool, user = self.get_tool_and_user(message)
+        update_tool_status(user, tool, TOOL_UPGRADED)
```
Comment (andyhd) [inline controlpanel/frontend/consumers.py]: I'm not keen on introducing new statuses - I don't think theres much value in differentiating between `Upgraded` and `Ready`.

File: controlpanel/frontend/jinja2/includes/forms.html [hunk 4]
```diff
@@ -56,8 +56,3 @@ <h2 class="heading-medium govuk-fieldset__heading">
     {{ caller() }}
   </fieldset>
 {%- endmacro %}
-
-{% macro delete_button(label, confirm="") -%}
-  <input type="submit" class="button button-warning right js-confirm" value="{{
-  label }}" data-confirm-message="{{ confirm | default("Are you sure?") }}">
-{%- endmacro %}
```
Comment (andyhd) [inline controlpanel/frontend/jinja2/includes/forms.html]: I thought this whole file was already deleted - why is this change here?

File: controlpanel/frontend/static/javascripts/modules/confirm.js [hunk 5]
```diff
@@ -16,7 +16,9 @@ moj.Modules.jsConfirm = {
       }
     });
 
-    $(document).on('click', `input[type="submit"].${this.confirmClass}`, (e) => {
+    // works on any children of a `<form>` with `confirmClass` but it's
+    // usually used on `<input type="submit">` or `<button>`
+    $(document).on('click', `form > .${this.confirmClass}`, (e) => {
```
Comment (andyhd) [inline controlpanel/frontend/static/javascripts/modules/confirm.js]: Not sure if this is meant to be in this PR?

File: controlpanel/frontend/static/javascripts/modules/tool-status.js [hunk 6]
```diff
@@ -38,10 +38,13 @@ moj.Modules.toolStatus = {
           break;
         case 'READY':
         case 'IDLED':
-          this.showActions(listener, ['open', 'restart', 'remove']);
+          this.showActions(listener, ['open', 'restart', 'upgrade', 'remove']);
+          break;
+        case 'UPGRADED':
+          this.showActions(listener, ['open']);
```
Comment (andyhd) [inline controlpanel/frontend/static/javascripts/modules/tool-status.js]: Why doesn't an upgraded tool have `restart`? (and `remove` - not that it exists yet)

File: controlpanel/frontend/consumers.py (hunk 3 above)
Comment (xoen) [reply controlpanel/frontend/consumers.py]: One good reason for being specific is that we know that we need to hide the "Upgrade" button once the tool is upgraded: Bear in mind we don't want to re-trigger a refresh and if we just recycle "ready" or "upgraded" we can't tell the difference.

File: controlpanel/frontend/jinja2/includes/forms.html (hunk 4 above)
Comment (xoen) [reply controlpanel/frontend/jinja2/includes/forms.html]: Maybe was deleted recently? It wasn't deleted when I made the change 😆

File: controlpanel/frontend/static/javascripts/modules/confirm.js (hunk 5 above)
Comment (xoen) [reply controlpanel/frontend/static/javascripts/modules/confirm.js]: I almost used this module and found out there was a little bug (not working with buttons) and fixed it as it was a small fix.

File: controlpanel/frontend/jinja2/includes/forms.html (hunk 4 above)
Comment (andyhd) [reply controlpanel/frontend/jinja2/includes/forms.html]: I must have it in a branch somewhere 😕

File: controlpanel/frontend/consumers.py (hunk 3 above)
Comment (andyhd) [reply controlpanel/frontend/consumers.py]: Ah, good point

File: controlpanel/frontend/static/javascripts/modules/tool-status.js (hunk 6 above)
Comment (xoen) [reply controlpanel/frontend/static/javascripts/modules/tool-status.js]: no `restart`: I imagine no point to restart something which was just upgraded - but I can add if you disagree
no `remove`: because it doesn't exist - although I can add if you think it's good 
 to leave for the future

Speaking of `restart`: does it make sense when something is already idled? I was tempted to remove but then I left as it was a bit out of scope and I thought there may have been a reason for it.

File: controlpanel/api/models/tool.py (hunk 2 above)
Comment (xoen) [reply controlpanel/api/models/tool.py]: I'm not keen on passing the `deployed_version` as I don't want to make this mandatory and when I make it optional the semantic of this gets weird: Does `deployed_version=None` means not deployed or out of date? I don't want to get into that as can lead to confusion/bugs.

But I like the idea of just having a `is_outdated` property which we set in the `ToolDeploymentManager` - it may not be the best solution but it would work for the time being.

File: controlpanel/frontend/static/javascripts/modules/tool-status.js (hunk 6 above)
Comment (andyhd) [reply controlpanel/frontend/static/javascripts/modules/tool-status.js]: They may want to restart if the tool fails to unidle.

Comment (xoen) [reply controlpanel/frontend/static/javascripts/modules/tool-status.js]: Wouldn't that be a separate state tho? It would be probably be "unknown" or "failed" no? (maybe I miss something).
